fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/AsyncLockRegistry.java

```diff
@@ -0,0 +1,161 @@
+         * backlog tasks are done and the current new read task can begin executing. The new read task can begin only after
+         * all the existing write tasks have been completed, to maintain exclusivity. However, the read task can begin
+         * concurrently with other read tasks that themselves should start after all write tasks.
+         */
+        private Pair<CompletableFuture<Void>, AsyncLock> withNewRead(CompletableFuture<Void> taskFuture) {
+            final var waitFuture = writes;
+            final var newReadFuture = CompletableFuture.allOf(reads, waitFuture.thenCompose(ignore -> taskFuture));
+            return ImmutablePair.of(waitFuture, new AsyncLock(newReadFuture, writes));
+        }
+
+        /**
+         * Constructs a new {@link AsyncLock} from the calling lock by stacking the new write future to its backlog tasks.
+         *
+         * @param taskFuture the {@link CompletableFuture} to complete when the lock has to be released.
+         *
+         * @return A pair of the new {@link AsyncLock} and the {@link CompletableFuture} that is completed when the
+         * backlog tasks are done and the current new write task can begin executing. The new write task can begin only after
+         * all the existing write tasks have been completed, to maintain exclusivity. It should also wait for one of more
+         * read tasks that can happen concurrently.
+         */
+        private Pair<CompletableFuture<Void>, AsyncLock> withNewWrite(CompletableFuture<Void> taskFuture) {
+            final var waitFuture = CompletableFuture.allOf(reads, writes);
+            final var newWriteFuture = waitFuture.thenCompose(ignore -> taskFuture);
+            return ImmutablePair.of(waitFuture, new AsyncLock(COMPLETED_FUTURE, newWriteFuture));
+        }
+    }
+
+    Map<LockIdentifier, AtomicReference<AsyncLock>> heldLocks = new ConcurrentHashMap<>();
+
+    public CompletableFuture<Void> getReadLock(@Nonnull LockIdentifier identifier, CompletableFuture<Void> taskFuture) {
+        return updateRefAndGetFuture(identifier, (parentLock) -> parentLock.withNewRead(taskFuture));
+    }
+
+    public CompletableFuture<Void> getWriteLock(@Nonnull LockIdentifier identifier, CompletableFuture<Void> taskFuture) {
+        return updateRefAndGetFuture(identifier, (parentLock) -> parentLock.withNewWrite(taskFuture));
+    }
+
+    private CompletableFuture<Void> updateRefAndGetFuture(@Nonnull LockIdentifier identifier, Function<AsyncLock, Pair<CompletableFuture<Void>, AsyncLock>> getNewLock) {
+        final var parentLockRef = heldLocks.computeIfAbsent(identifier, ignore -> new AtomicReference<>(NO_LOCKING_LOCK));
+        AsyncLock parentLock;
+        Pair<CompletableFuture<Void>, AsyncLock> waitFutureAndNewLock;
+        do {
+            parentLock = parentLockRef.get();
+            waitFutureAndNewLock = getNewLock.apply(parentLock);
+        } while (!parentLockRef.compareAndSet(parentLock, waitFutureAndNewLock.getRight()));
```

Comment:
I think it would need to be `updateAndGet`. I don't actually think this would reduce thrashing, as the logic within `updateAndGet` is similar to the logic here (in that it calls the update lambda and then attempts to change the atomic reference's value if and only if the reference hasn't been updated). It may be slightly better to rely on the library method from a code design method, but only marginally so

Reply:
done